### DISCUSSION TITLE
#30 Horizon

### DISCUSSION THREAD

**Discussion by GaganPrasad** [cssbattle/forum#168] [Q&A] [answered]
Any suggestion on how to optimize this code any further?

`<p style=box-shadow:0+0+0+45q#F7EC7D,0+0+0+23vw#E38F66,0+0+0+2in#AA445F,0+0+0+5in#62306D>`

**Comment by RazvanManolache** [comment]
You reached the same count I did, but I did it with gradient. I remember I tried at some point to use only #F7EC7D and #62306D and then to put a transparent color on top of them. I think that's when I tried my hand making a js to do the math for me. Didn't work for me. But if you find the magic color, then yeah, maybe that.

**Comment by SelenIT** [comment, marked as answer]
Yes, multiple box-shadows and 4-hex-digit semi-transparent color substitutions is the key. By the way, great solution with the `p` element! However... maybe it's another way without that many zeroes in the `box-shadow` value?..

**Reply by GaganPrasad** [reply]
`<style>*>*{background:linear-gradient(#ed08 53q,#c409 0 106q,#802b 0 212q,#415d 0`
This seems exactly the same but results to give 0%
![Screenshot 2021-11-15 at 10 47 49 AM](https://user-images.githubusercontent.com/49602219/141726637-2b8c46fd-0f74-4113-82fe-c947b4558028.png)

and with box-shadow 4-hex-digit code colors are way off, resulting in 0.3% for the combinations I tried 🥲 .
Used this link https://bittersmann.de/cssbattle/color/three to get the 4-hex-digit code

**Reply by siva-kannan3** [reply]
@SelenIT can we possibly reduce the many zeros in the box shadow by setting spread radius alone?

**Reply by SelenIT** [reply]
@GaganPrasad colors don't have to be mixed with default white background only... 😉

@siva-kannan3 unfortunately, no. Per current CSS standards (implemented in browsers), spread radius is optional, but offsets are mandatory for `box-shadow`.

**Reply by GaganPrasad** [reply]
@SelenIT can you please explain a little more I don't exactly understand how it works,

Like if this is the code 
`<p><style>p{width:100;height:100;background:#FFC0CB;box-shadow:0+0+0+10px #a020f0`

and I need to find the transparent code so will mix #FFC0CB and #a020f0 and as box-shadow color #a020f0 will be at bottom we get #fbc, and this will take the place of the background color #FFC0CB. right? or I am understanding wrong?

**Reply by SelenIT** [reply]
@GaganPrasad first, I have to apologize that, specifically in Target 30, the alpha-color optimization stopped to work after one of the backend updates, so apparently 82-char-solution is not possible now, only 84 chars without that trick 😳 I totally forgot about it.

That said, this trick is still useful for other targets. It can probably better be illustrated with [examples 11 and 12 from @alexzaworski's article on css-tricks](https://css-tricks.com/a-css-golfing-exercise/#attempt-11-117-bytes) (for Target 7).

In that example, we have a medium green shadow (`#1A4341`) on the dark green viewport background (`#0B2429`). We look for a semi-transparent color that, being overlaid over that background, will give us `#1A4341` (or something really close to it so the difference can be tolerated by the scoring algorithm).

With [simple alpha compositing](https://drafts.fxtf.org/compositing-1/#simplealphacompositing), given that the opacity of the new color is `α` (between 0 and 1), each channel (R, G, B) of the resulting color gets `α` part of the respective channel of the new color and `1 - α` part of that channel of the background color. For the example from the article:

```
background = #0B2429 = rgb(11, 36, 41)
new = #7fd2 = #77ffdd22 = rgba(119, 255, 221, 34/255) = rgba(119, 255, 221, 0.1333)
result = rgba(
  119 * 0.1333 + 11 * (1 - 0.1333),
  255 * 0.1333 + 36 * (1 - 0.1333),
  221 * 0.1333 + 41 * (1 - 0.1333) 
) = rgba(119*0.1333 + 11*0.8667, 255*0.1333 + 36*0.8667, 221*0.113 + 41*0.8667) = rgba(25, 65, 65) = #194141
```
which differs from the target color `#1A4341` by only 1 in the red channel and by 2 in the green channel.

Since we need only colors that can be written in the 4-digit form, we can simply iterate over them all with a nested loop. I tried to write [a script for that](https://codepen.io/SelenIT/pen/JaeYpM?editors=0011) (currently it outputs candidate colors into console sorted by "quadratic error", I'm goind to add a better UI to it). From my experience, colors with error <= 2 work most of the time, 
colors with 2 < error < 4 work occasionally, and colors with larger error almost never work. But sometimes surplising things happen like seemingly "worse" color works and seemingly "better" one doesn't 😃. Also, sometimes it's possible to adjust the resulting color a bit by tweaking the last digit of one or two channels of the background.

**Reply by GaganPrasad** [reply]
Wow, thank you.  ❤️
